packages/ui/src/shapes_ui.rs

```diff
@@ -0,0 +1,339 @@
+        if fill_color == Color::White {
+            fill_color = APP_STATE.write().get_fill_color_and_advance();
+        }
+        let style = Style::new(fill_color);
+        *CANVAS_TRACKER.write() = Some(Rc::new(NewRectTracker::new(evt, shape_id, style)))
+    };
+
+    rsx! {
+        rect {
+            id: "background",
+            x: "0",
+            y: "0",
+            width: "100%",
+            height: "100%",
+            fill: "white",
+            onmousedown: canvas_mouse_down
+        }
+    }
+}
+
+#[component]
+fn RenderedShapes() -> Element {
+    let doc: &Document = &*DOC.read();
+    // FIXME: It would be better if this iterator lived in the document
+    let rendered_shapes_iter =
+        doc.get_sequence()
+            .iter()
+            .filter_map(|shape_id| match &doc.get_shape_by_id(*shape_id) {
+                Some(shape) => Some(render_shape(*shape_id, shape)),
+                None => None,
+            });
+    rsx! {
+        for shape in rendered_shapes_iter {
+            { shape }
+        }
+    }
+}
+
+/* Shape dragging */
+
+struct ShapeDragTracker {
+    mouse_down_x: f64,
+    mouse_down_y: f64,
+    shape_id: ShapeId,
+    initial_geometry: Geometry,
+}
+
+impl ShapeDragTracker {
+    fn new(mouse_down: MouseEvent, shape_id: ShapeId, initial_geometry: &Geometry) -> Self {
+        let mouse_down_x = mouse_down.data.page_coordinates().x;
+        let mouse_down_y = mouse_down.data.page_coordinates().y;
+        Self {
+            mouse_down_x,
+            mouse_down_y,
+            shape_id,
+            initial_geometry: initial_geometry.clone(),
+        }
+    }
+
+    fn update_shape_for_drag(&self, mouse: &MouseEvent) {
+        let event_x = mouse.data.page_coordinates().x;
+        let event_y = mouse.data.page_coordinates().y;
+        let new_geometry = self
+            .initial_geometry
+            .offset_by(event_x - self.mouse_down_x, event_y - self.mouse_down_y);
+        DOC.write()
+            .update_geometry_for_shape_id(&self.shape_id, new_geometry)
+    }
+}
+
+impl Tracker for ShapeDragTracker {
+    fn track_mouse_move(&self, evt: &MouseEvent) -> TrackerNext {
+        self.update_shape_for_drag(evt);
+        TrackerNext::Continue
+    }
+    fn track_mouse_up(&self, evt: &MouseEvent) {
+        self.update_shape_for_drag(evt)
+    }
+}
+
+/* SVG generation  */
+
+fn svg_color(color: &Color) -> String {
+    match color {
+        Color::Red => "red".to_string(),
+        Color::Orange => "orange".to_string(),
+        Color::Yellow => "yellow".to_string(),
+        Color::Green => "green".to_string(),
+        Color::Blue => "blue".to_string(),
+        Color::Indigo => "indigo".to_string(),
+        Color::Violet => "violet".to_string(),
+        Color::White => "white".to_string(),
+        Color::Black => "black".to_string(),
+    }
+}
+
+fn render_shape(shape_id: ShapeId, shape: &Shape) -> Element {
+    let id_string = format!("shape_{}", shape_id);
```

Comment:
```suggestion
    let id_string = format!("shape_{shape_id}");
```

Use Clippy; it will automatically take care of trivia like this.